### PR TITLE
Add logic to shorten paths when they exceed 259 chars

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes: #1513

Add logic to try to shorten paths when they exceed 259 characters.

Also modified a test case to trigger this case.

### CODE REVIEW

**Comment by JohnTortugo** [comment]
PTAL; https://github.com/aspnet/AspNetCore/pull/4005 is blocked on this.

**Comment by jcagme** [comment]
How do you shorten the path by keep pointing to the same file? Is there a way to "move" the actual file to a location with a short path, sign it there and replace the original? Also, who dictates the max path lenght of 260, MicroBuild?

**Comment by JohnTortugo** [comment]
> How do you shorten the path by keep pointing to the same file? Is there a way to "move" the actual file to a location with a short path, sign it there and replace the original? Also, who dictates the max path lenght of 260, MicroBuild?

The problem here is the path where we extract container files might be too long, this PR shortens the path where the files will be extracted. The original code extract the container to a folder named after the hash string of the file, something like `..\ContainerSigning\16D285869A01B4C54E3C5CDA81CE0A9B075BD2A43F824EAFB00AFB3B3FC4497F\..`. A hash that long isn't necessary in most situations. The logic I added tries different ways to shorten that.

The max path length is a restriction [imposed by the filesystem](https://docs.microsoft.com/en-us/windows/desktop/fileio/naming-a-file).

**Comment by jcagme** [comment]
I thought it would be the same as PATH in batch.

**Comment by JohnTortugo** [comment]
PTAL; pushed with the changes you suggested.

**Comment by tmat** [comment]
This is not the change I suggested. I meant to assign each `FileSignInfo` added to `_filesByContentKey` a number (0, 1, 2, 3, ...), could be `_filesByContentKey.Count` when the entry is being added and then use the number as the name of the subdirectory instead of the hash.

**Comment by JohnTortugo** [comment]
@tmat I think I addressed all your comments. Is there anything else blocking? Thanks

**Comment by JohnTortugo** [comment]
Planning on merging this by end of the day if no one objects.

**Comment by tmat** [comment]
👍

**Comment by nguerrera** [comment]
We're hitting this in dotnet/sdk official builds and I believe we have latest arcade package. At least we have the one that was last pushed to us? When does this get pushed out?

**Comment by JohnTortugo** [comment]
> We're hitting this in dotnet/sdk official builds and I believe we have latest arcade package. At least we have the one that was last pushed to us. When does this get pushed out?

You are using the default version that comes with the SDK and since the SDK doesn't use the latest version of the pacakge you're not getting the patch. I submitted a PR to update the default version on the SDK #1598 . In the meantime you can add this to your `eng\Version.props` to use a version that has the patch:

`<MicrosoftDotNetSignToolVersion>1.0.0-beta.18614.3</MicrosoftDotNetSignToolVersion>`

**Comment by nguerrera** [comment]
I see, thanks. I did not realize that it worked like that. Good to know.

**Comment by nguerrera** [comment]
I will wait for the update of the default version so that we don't have to maintain the version in dotnet/sdk.

File: src/Microsoft.DotNet.SignTool/src/Configuration.cs [hunk 1]
```diff
@@ -463,7 +463,8 @@ private bool TryBuildZipData(FileSignInfo zipFileSignInfo, out ZipData zipData)
                         var fileName = Path.GetFileName(relativePath);
                         if (!_filesByContentKey.TryGetValue(new SignedFileContentKey(contentHash, fileName), out var fileSignInfo))
                         {
-                            string tempPath = Path.Combine(_pathToContainerUnpackingDirectory, ContentUtil.HashToString(contentHash), relativePath);
+                            string tempPath = Path.Combine(_pathToContainerUnpackingDirectory, _filesByContentKey.Count().ToString(), relativePath);
```
Comment (weshaggard) [inline src/Microsoft.DotNet.SignTool/src/Configuration.cs]: There are some cases (namely the one I'm working on) that rely on the content hash being included in the file path so as part of my changes I going to make using the content has as an option.

Comment (tmat) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs]: What are those cases?

Comment (weshaggard) [reply src/Microsoft.DotNet.SignTool/src/Configuration.cs]: See https://github.com/dotnet/arcade/pull/1606.

**Pull request merged**